File: packages/crypto/src/otp-migration.ts
import { parseOtpauth, type TOTPParams, type OTPAlgorithm } from "./totp.js";

/**
 * Importing existing 2FA accounts on the web — the TS port of the iOS
 * `AuthBoxCrypto/OTPImport.swift`.
 *
 * Scan, import, and migrate reduce to one problem: parse a string into
 * `TOTPParams[]`. Handles a Google Authenticator `otpauth-migration://` export
 * (base64 + a minimal dependency-free protobuf reader), a single `otpauth://`
 * URI, or a newline-separated list. Invalid entries are skipped, never fatal.
 */
export function parseOTPImport(raw: string): TOTPParams[] {
  const trimmed = raw.trim();
  if (trimmed.length === 0) return [];

  if (trimmed.toLowerCase().startsWith("otpauth-migration://")) {
    return parseMigration(trimmed);
  }
  return trimmed
    .split(/\r?\n/)
    .map((line) => parseOtpauth(line.trim()))
    .filter((p): p is TOTPParams => p !== null);
}

/** Decode a Google Authenticator `otpauth-migration://offline?data=...` URI. */
export function parseMigration(raw: string): TOTPParams[] {
  // Extract `data` from the raw query manually: URLSearchParams would turn a
  // literal '+' in the base64 into a space (form-encoding semantics).
  const query = raw.split("?")[1] ?? "";
  const dataKv = query.split("&").find((kv) => kv.startsWith("data="));
  if (!dataKv) return [];

  let dataParam: string;
  try {
    dataParam = decodeURIComponent(dataKv.slice("data=".length));
  } catch {
    return [];
  }

  const bytes = base64ToBytes(dataParam);
  if (!bytes) return [];
  return decodeMigrationPayload(bytes);
}

function base64ToBytes(input: string): Uint8Array | null {
  try {
    let clean = input.replace(/\s/g, "");
    const remainder = clean.length % 4;
    if (remainder > 0) clean += "=".repeat(4 - remainder);
    const bin = atob(clean);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
  } catch {
    return null;
  }
}

// --- Protobuf decode (otpauth-migration MigrationPayload) ---

/** `MigrationPayload { repeated OtpParameters otp_parameters = 1; ... }` */
function decodeMigrationPayload(data: Uint8Array): TOTPParams[] {
  const reader = new ProtobufReader(data);
  const out: TOTPParams[] = [];
  let field = reader.nextField();
  while (field !== null) {
    if (field.number === 1 && field.bytes) {
      const params = decodeOtpParameters(field.bytes);
      if (params) out.push(params);
    }
    field = reader.nextField();
  }
  return out;
}

/**
 * `OtpParameters { bytes secret = 1; string name = 2; string issuer = 3;`
 * ` Algorithm algorithm = 4; DigitCount digits = 5; OtpType type = 6; }`
 *
 * Algorithm: 1 SHA1, 2 SHA256, 3 SHA512, 4 MD5 (unsupported → skip).
 * DigitCount: 1 six, 2 eight. OtpType: 1 HOTP (skip — TOTP is time-based), 2 TOTP.
 */
function decodeOtpParameters(data: Uint8Array): TOTPParams | null {
  const reader = new ProtobufReader(data);
  const decoder = new TextDecoder();
  let secret: Uint8Array = new Uint8Array(0);
  let name: string | undefined;
  let issuer: string | undefined;
  let algorithm: OTPAlgorithm = "SHA1";
  let digits = 6;
  let isTotp = true;

  let field = reader.nextField();
  while (field !== null) {
    if (field.number === 1 && field.bytes) {
      secret = field.bytes;
    } else if (field.number === 2 && field.bytes) {
      name = decoder.decode(field.bytes);
    } else if (field.number === 3 && field.bytes) {
      issuer = decoder.decode(field.bytes);
    } else if (field.number === 4 && field.varint !== undefined) {
      const v = Number(field.varint);
      if (v === 1) algorithm = "SHA1";
      else if (v === 2) algorithm = "SHA256";
      else if (v === 3) algorithm = "SHA512";
      else return null; // MD5 / unspecified — unsupported, skip the account
    } else if (field.number === 5 && field.varint !== undefined) {
      digits = Number(field.varint) === 2 ? 8 : 6;
    } else if (field.number === 6 && field.varint !== undefined) {
      isTotp = Number(field.varint) !== 1; // 1 = HOTP
    }
    field = reader.nextField();
  }

  if (secret.length === 0 || !isTotp) return null;
  // The migration payload carries no period; Google Authenticator uses 30s.
  return {
    secret,
    digits,
    period: 30,
    algorithm,
    issuer: issuer && issuer.length > 0 ? issuer : undefined,
    account: name,
  };
}

interface ProtobufField {
  number: number;
  wireType: number;
  varint?: bigint;
  bytes?: Uint8Array;
}

/**
 * Minimal protobuf wire-format reader — only the two wire types the migration
 * payload uses (varint, length-delimited). 32-/64-bit fields are skipped and
 * unknown wire types stop the read, so newer exports degrade gracefully.
 */
class ProtobufReader {
  private readonly data: Uint8Array;
  private index = 0;

  constructor(data: Uint8Array) {
    this.data = data;
  }

  nextField(): ProtobufField | null {
    const key = this.readVarint();
    if (key === null) return null;
    const number = Number(key >> 3n);
    const wireType = Number(key & 0x07n);

    switch (wireType) {
      case 0: {
        const v = this.readVarint();
        return v === null ? null : { number, wireType, varint: v };
      }
      case 2: {
        const len = this.readVarint();
        if (len === null) return null;
        const bytes = this.read(Number(len));
        return bytes === null ? null : { number, wireType, bytes };
      }
      case 5:
        return this.read(4) === null ? null : this.nextField();
      case 1:
        return this.read(8) === null ? null : this.nextField();
      default:
        return null;
    }
  }

  private readVarint(): bigint | null {
    let result = 0n;
    let shift = 0n;
    while (this.index < this.data.length) {
      const byte = this.data[this.index];
      this.index += 1;
      result |= BigInt(byte & 0x7f) << shift;
      if ((byte & 0x80) === 0) return result;
      shift += 7n;
      if (shift >= 64n) return null;
    }
    return null;
  }

  private read(n: number): Uint8Array | null {
    if (n < 0 || this.index + n > this.data.length) return null;
    const sub = this.data.subarray(this.index, this.index + n);
    this.index += n;
    return sub;
  }
}

File: apps/ios/AuthBoxCrypto/Sources/AuthBoxCrypto/OTPMigration.swift
import Foundation

/// Importing existing 2FA accounts into the vault.
///
/// Two ingestion surfaces, one output type (`[TOTP]`):
///  - `otpauth-migration://offline?data=...` — Google Authenticator's "Export
///    accounts" / "Transfer accounts" QR. The `data` param is base64 of a
///    protobuf `MigrationPayload` that carries many accounts in one code.
///  - One or more `otpauth://totp/...` URIs (or bare base32 secrets), one per
///    line — the single-account QR most services show, or a pasted list.
///
/// The migration protobuf is decoded by a minimal, dependency-free wire-format
/// reader (no SwiftProtobuf) — the same "implement the standard by hand, keep it
/// testable" posture as `TOTP` itself.
public enum OTPImport {

    /// Parse any scanned/pasted authenticator string into the TOTP accounts it
    /// carries. Handles a migration URI, a single `otpauth://` URI, or a
    /// newline-separated list. Invalid entries are skipped, never fatal.
    public static func parse(_ raw: String) -> [TOTP] {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        if trimmed.lowercased().hasPrefix("otpauth-migration://") {
            return parseMigration(trimmed)
        }
        // One-or-many otpauth:// URIs / bare secrets, split on line breaks.
        return trimmed
            .split(whereSeparator: \.isNewline)
            .compactMap { TOTP.parse(String($0)) }
    }

    /// Decode a Google Authenticator `otpauth-migration://offline?data=...` URI
    /// into the accounts it carries.
    public static func parseMigration(_ raw: String) -> [TOTP] {
        guard let comps = URLComponents(string: raw),
              let dataParam = (comps.queryItems ?? [])
                  .first(where: { $0.name.lowercased() == "data" })?.value
        else { return [] }

        // URLComponents already percent-decodes the value; Google's payload is
        // standard base64 (may contain '+' and '/'). Try direct, then a cleaned
        // pass that tolerates whitespace and missing padding.
        guard let payload = base64Decode(dataParam) else { return [] }
        return decodeMigrationPayload(payload)
    }

    private static func base64Decode(_ s: String) -> Data? {
        if let d = Data(base64Encoded: s) { return d }
        var cleaned = s.replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "\n", with: "")
        // Pad to a multiple of 4 so unpadded payloads still decode.
        let remainder = cleaned.count % 4
        if remainder > 0 { cleaned += String(repeating: "=", count: 4 - remainder) }
        return Data(base64Encoded: cleaned)
    }

    // MARK: - Protobuf decode (otpauth-migration MigrationPayload)

    /// `MigrationPayload { repeated OtpParameters otp_parameters = 1; ... }`
    static func decodeMigrationPayload(_ data: Data) -> [TOTP] {
        var reader = ProtobufReader(data)
        var totps: [TOTP] = []
        while let field = reader.nextField() {
            // Field 1 = repeated OtpParameters (length-delimited).
            if field.number == 1, case let .lengthDelimited(bytes) = field.value {
                if let totp = decodeOtpParameters(bytes) { totps.append(totp) }
            }
        }
        return totps
    }

    /// `OtpParameters { bytes secret = 1; string name = 2; string issuer = 3;`
    /// ` Algorithm algorithm = 4; DigitCount digits = 5; OtpType type = 6; }`
    ///
    /// Algorithm: 1 SHA1, 2 SHA256, 3 SHA512, 4 MD5 (unsupported → skip).
    /// DigitCount: 1 six, 2 eight. OtpType: 1 HOTP (skip — TOTP is time-based), 2 TOTP.
    private static func decodeOtpParameters(_ data: Data) -> TOTP? {
        var reader = ProtobufReader(data)
        var secret = Data()
        var name: String?
        var issuer: String?
        var algorithm: TOTP.Algorithm = .sha1
        var digits = 6
        var isTotp = true

        while let field = reader.nextField() {
            switch (field.number, field.value) {
            case (1, .lengthDelimited(let b)): secret = b
            case (2, .lengthDelimited(let b)): name = String(data: b, encoding: .utf8)
            case (3, .lengthDelimited(let b)): issuer = String(data: b, encoding: .utf8)
            case (4, .varint(let v)):
                switch v {
                case 1: algorithm = .sha1
                case 2: algorithm = .sha256
                case 3: algorithm = .sha512
                default: return nil   // MD5 / unspecified — not supported, skip the account
                }
            case (5, .varint(let v)): digits = (v == 2) ? 8 : 6
            case (6, .varint(let v)): isTotp = (v != 1)   // 1 = HOTP
            default: break
            }
        }

        guard !secret.isEmpty, isTotp else { return nil }
        // The migration payload carries no period; Google Authenticator uses 30s.
        return TOTP(
            secret: secret,
            digits: digits,
            period: 30,
            algorithm: algorithm,
            issuer: (issuer?.isEmpty == false) ? issuer : nil,
            account: name
        )
    }
}

/// Minimal protobuf wire-format reader — only the two wire types the migration
/// payload uses: varint (0) and length-delimited (2). 32-/64-bit fields are
/// skipped and unknown wire types stop the read, so newer export versions
/// degrade gracefully instead of crashing.
struct ProtobufReader {
    enum Value: Equatable {
        case varint(UInt64)
        case lengthDelimited(Data)
    }
    struct Field { let number: Int; let value: Value }

    private let data: Data
    private var index: Int

    /// Re-base into a contiguous, zero-indexed buffer so `Data` slices (whose
    /// `startIndex` may be non-zero) can't corrupt offset math.
    init(_ data: Data) {
        self.data = Data(data)
        self.index = 0
    }

    mutating func nextField() -> Field? {
        guard let key = readVarint() else { return nil }
        let number = Int(key >> 3)
        let wireType = Int(key & 0x07)
        switch wireType {
        case 0:                                   // varint
            guard let v = readVarint() else { return nil }
            return Field(number: number, value: .varint(v))
        case 2:                                   // length-delimited
            guard let len = readVarint(), let bytes = read(Int(len)) else { return nil }
            return Field(number: number, value: .lengthDelimited(bytes))
        case 5:                                   // 32-bit fixed — skip
            guard read(4) != nil else { return nil }
            return nextField()
        case 1:                                   // 64-bit fixed — skip
            guard read(8) != nil else { return nil }
            return nextField()
        default:                                  // group / unknown — stop
            return nil
        }
    }

    private mutating func readVarint() -> UInt64? {
        var result: UInt64 = 0
        var shift: UInt64 = 0
        while index < data.count {
            let byte = data[index]
            index += 1
            result |= UInt64(byte & 0x7f) << shift
            if byte & 0x80 == 0 { return result }
            shift += 7
            if shift >= 64 { return nil }
        }
        return nil
    }

    private mutating func read(_ n: Int) -> Data? {
        guard n >= 0, index + n <= data.count else { return nil }
        let sub = data.subdata(in: index..<(index + n))
        index += n
        return sub
    }
}
